File: tests/selenium-python/src/pages/comms_flow_app.py
from selenium.webdriver.common.by import By
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import Select, WebDriverWait


class CommsFlowApp:
    def __init__(self, driver, base_url):
        self.driver = driver
        self.base_url = base_url
        self.wait = WebDriverWait(driver, 10)

    def open(self):
        self.driver.get(self.base_url)
        self.wait_for_test_id("login-page")
        return self

    def reset_state(self):
        self.driver.get(self.base_url)
        self.driver.execute_script("localStorage.clear();")
        self.driver.refresh()
        self.wait_for_test_id("login-page")
        return self

    def login_as(self, username, password="commsflow123"):
        self.type_by_test_id("username-input", username)
        self.type_by_test_id("password-input", password)
        self.click_by_test_id("login-submit")
        return self

    def open_protected_archive_while_logged_out(self):
        self.driver.get(f"{self.base_url}/archive")
        self.wait_for_test_id("login-page")
        return self

    def expect_invalid_login_error(self):
        assert "Invalid credentials" in self.find_by_test_id("login-error").text
        assert self.find_by_test_id("login-page").is_displayed()
        return self

    def expect_manager_lands_on_dashboard(self):
        self.wait_for_test_id("dashboard-page")
        assert "Comms Manager" in self.find_by_test_id("current-user").text
        return self

    def expect_reviewer_lands_on_templates(self):
        self.wait_for_test_id("templates-page")
        assert "Compliance Reviewer" in self.find_by_test_id("current-user").text
        return self

    def expect_session_persists_after_refresh(self):
        self.driver.refresh()
        self.wait_for_test_id("dashboard-page")
        assert "Comms Manager" in self.find_by_test_id("current-user").text
        return self

    def expect_manager_access_profile(self):
        assert "Comms Manager" in self.find_by_test_id("access-profile").text
        permissions = self.find_by_test_id("permission-list").text
        assert "Create templates" in permissions
        assert "Submit for approval" in permissions
        assert "Send campaigns" in permissions
        assert "3 permissions" in self.find_by_test_id("current-permissions").text
        return self

    def expect_reviewer_access_profile(self):
        self.open_dashboard()
        assert "Compliance Reviewer" in self.find_by_test_id("access-profile").text
        permissions = self.find_by_test_id("permission-list").text
        assert "Approve regulated notices" in permissions
        assert "1 permissions" in self.find_by_test_id("current-permissions").text
        return self

    def logout(self):
        self.click_by_test_id("logout-button")
        self.wait_for_test_id("login-page")
        return self

    def expect_topbar_navigation(self):
        self.open_dashboard()
        self.open_customers()
        self.open_templates()
        self.open_campaigns()
        self.open_inbox()
        self.open_archive()
        return self

    def open_templates(self):
        self.click_by_test_id("nav-templates")
        self.wait_for_test_id("templates-page")
        return self

    def create_template(self, name="Policy Renewal Notice"):
        self.type_by_test_id("template-name-input", name)
        self.click_by_test_id("template-create-button")
        self.wait_for_test_id("template-card-policy-renewal-notice")
        return self

    def expect_template_draft_metadata(self):
        card = self.find_by_test_id("template-card-policy-renewal-notice").text
        assert "version 1" in card.lower()
        assert "Owner: Comms Manager" in card
        assert "Draft" in self.find_by_test_id("template-status-policy-renewal-notice").text
        return self

    def expect_duplicate_template_is_ignored(self):
        self.create_template()
        assert self.count_test_id("template-card-policy-renewal-notice") == 1
        return self

    def expect_templates_empty_state(self):
        assert "No communication templates" in self.find_by_test_id("templates-empty-state").text
        return self

    def expect_template_details(self):
        self.click_by_test_id("template-details-policy-renewal-notice")
        details = self.find_by_test_id("template-details-policy-renewal-notice").text
        assert "Compliance approval required" in details
        assert "Email, SMS, Portal, Print" in details
        return self

    def submit_template_for_approval(self):
        self.click_by_test_id("submit-template-policy-renewal-notice")
        self.wait_for_text("template-status-policy-renewal-notice", "Pending Approval")
        return self

    def expect_manager_cannot_approve_pending_template(self):
        assert "Approval is restricted" in self.find_by_test_id("manager-approval-blocked-policy-renewal-notice").text
        assert not self.has_test_id("approve-template-policy-renewal-notice")
        return self

    def expect_reviewer_cannot_create_template(self):
        assert "Template creation restricted" in self.find_by_test_id("template-permission-note").text
        assert not self.has_test_id("template-form")
        return self

    def approve_template(self):
        self.click_by_test_id("approve-template-policy-renewal-notice")
        self.wait_for_text("template-status-policy-renewal-notice", "Approved")
        return self

    def expect_approve_button_is_hidden_after_approval(self):
        assert not self.has_test_id("approve-template-policy-renewal-notice")
        return self

    def open_campaigns(self):
        self.click_by_test_id("nav-campaigns")
        self.wait_for_test_id("campaigns-page")
        return self

    def expect_campaign_requires_approved_template(self):
        assert "approved template is required" in self.find_by_test_id("campaign-no-template-warning").text
        assert self.find_by_test_id("campaign-send-button").get_attribute("disabled") is not None
        return self

    def expect_reviewer_cannot_send_campaign(self):
        assert "Only Comms Managers can send campaigns" in self.find_by_test_id("campaign-permission-warning").text
        assert self.find_by_test_id("campaign-send-button").get_attribute("disabled") is not None
        return self

    def expect_campaigns_empty_state(self):
        assert "No campaigns" in self.find_by_test_id("campaigns-empty-state").text
        return self

    def send_policy_renewal_campaign(self):
        self.send_policy_renewal_campaign_to_customers(["cust-001", "cust-002", "cust-003", "cust-004"])
        self.expect_campaign_recipients(4)
        return self

    def send_policy_renewal_campaign_to_customers(self, customer_ids):
        Select(self.find_by_test_id("campaign-template-select")).select_by_value("policy-renewal-notice")
        for customer_id in ["cust-001", "cust-002", "cust-003", "cust-004"]:
            checkbox = self.find_by_test_id(f"customer-checkbox-{customer_id}")
            should_be_checked = customer_id in customer_ids
            if checkbox.is_selected() != should_be_checked:
                checkbox.click()
        self.click_by_test_id("campaign-send-button")
        self.wait_for_test_id("campaign-card-policy-renewal-may-2026")
        self.wait_for_text("campaign-status-policy-renewal-may-2026", "Sent")
        return self

    def attempt_campaign_without_recipients(self):
        Select(self.find_by_test_id("campaign-template-select")).select_by_value("policy-renewal-notice")
        for customer_id in ["cust-001", "cust-002", "cust-003", "cust-004"]:
            checkbox = self.find_by_test_id(f"customer-checkbox-{customer_id}")
            if checkbox.is_selected():
                checkbox.click()
        self.click_by_test_id("campaign-send-button")
        self.expect_campaigns_empty_state()
        return self

    def expect_campaign_recipients(self, count):
        assert f"Recipients: {count}" in self.find_by_test_id("campaign-card-policy-renewal-may-2026").text
        return self

    def expect_campaign_details(self):
        self.click_by_test_id("campaign-details-policy-renewal-may-2026")
        details = self.find_by_test_id("campaign-details-policy-renewal-may-2026").text
        assert "Customer preference based routing" in details
        assert "Archive records" in details
        return self

    def open_archive(self):
        self.click_by_test_id("nav-archive")
        self.wait_for_test_id("archive-page")
        return self

    def expect_archive_evidence(self):
        assert "Email" in self.find_by_test_id("archive-record-avery-brooks").text
        assert "SMS" in self.find_by_test_id("archive-record-maya-chen").text
        assert "Portal" in self.find_by_test_id("archive-record-nora-singh").text
        assert "Print" in self.find_by_test_id("archive-record-elliot-ward").text
        return self

    def expect_archive_evidence_for_selected_customers(self):
        assert "Email" in self.find_by_test_id("archive-record-avery-brooks").text
        assert "SMS" in self.find_by_test_id("archive-record-maya-chen").text
        assert not self.has_test_id("archive-record-nora-singh")
        assert not self.has_test_id("archive-record-elliot-ward")
        return self

    def expect_archive_initial_empty_state(self):
        assert "No communication evidence" in self.find_by_test_id("archive-initial-empty-state").text
        return self

    def expect_archive_search_works(self):
        self.type_by_test_id("archive-search-input", "Maya")
        assert self.find_by_test_id("archive-record-maya-chen").is_displayed()
        assert not self.has_test_id("archive-record-avery-brooks")
        self.type_by_test_id("archive-search-input", "FAX")
        assert "No archive records" in self.find_by_test_id("archive-empty-state").text
        return self

    def expect_archive_search_by_policy_number(self):
        self.type_by_test_id("archive-search-input", "POL-33810")
        assert "Portal" in self.find_by_test_id("archive-record-nora-singh").text
        assert not self.has_test_id("archive-record-maya-chen")
        return self

    def expect_archive_search_by_campaign_channel_status_and_trimmed_case(self):
        self.type_by_test_id("archive-search-input", "policy renewal may")
        assert self.find_by_test_id("archive-record-avery-brooks").is_displayed()
        assert self.find_by_test_id("archive-record-elliot-ward").is_displayed()

        self.type_by_test_id("archive-search-input", "  portal  ")
        assert "Portal" in self.find_by_test_id("archive-record-nora-singh").text
        assert not self.has_test_id("archive-record-maya-chen")
        assert "Portal: 1" in self.find_by_test_id("channel-summary").text

        self.type_by_test_id("archive-search-input", "sent")
        summary = self.find_by_test_id("channel-summary").text
        assert "Email: 1" in summary
        assert "SMS: 1" in summary
        assert "Portal: 1" in summary
        assert "Print: 1" in summary

        self.type_by_test_id("archive-search-input", "")
        self.expect_archive_evidence()
        return self

    def open_dashboard(self):
        self.click_by_test_id("nav-dashboard")
        self.wait_for_test_id("dashboard-page")
        return self

    def expect_dashboard_metrics(self, pending, approved, sent, archive):
        assert str(pending) in self.find_by_test_id("metric-pending-templates").text
        assert str(approved) in self.find_by_test_id("metric-approved-templates").text
        assert str(sent) in self.find_by_test_id("metric-sent-campaigns").text
        assert str(archive) in self.find_by_test_id("metric-archive-records").text
        return self

    def expect_dashboard_evidence(self):
        assert "1" in self.find_by_test_id("metric-approved-templates").text
        assert "1" in self.find_by_test_id("metric-sent-campaigns").text
        assert "4" in self.find_by_test_id("metric-archive-records").text
        assert "Campaign Policy Renewal May 2026 sent to 4 customers" in self.find_by_test_id("audit-trail").text
        return self

    def expect_audit_trail_contains_workflow_in_newest_first_order(self):
        audit = self.find_by_test_id("audit-trail").text
        assert "Comms Manager signed in" in audit
        assert "Template Policy Renewal Notice created as draft" in audit
        assert "Template Policy Renewal Notice moved to Pending Approval" in audit
        assert "Template Policy Renewal Notice moved to Approved" in audit
        assert "Campaign Policy Renewal May 2026 sent to 4 customers" in audit
        assert audit.index("Campaign Policy Renewal May 2026 sent to 4 customers") < audit.index("Template Policy Renewal Notice moved to Approved")
        assert audit.index("Template Policy Renewal Notice moved to Approved") < audit.index("Template Policy Renewal Notice moved to Pending Approval")
        return self

    def open_customers(self):
        self.click_by_test_id("nav-customers")
        self.wait_for_test_id("customers-page")
        return self

    def expect_customers_page_data(self):
        assert "Avery Brooks" in self.find_by_test_id("customer-row-cust-001").text
        assert "POL-10491" in self.find_by_test_id("customer-row-cust-001").text
        assert "Email" in self.find_by_test_id("customer-row-cust-001").text
        assert "SMS" in self.find_by_test_id("customer-row-cust-002").text
        assert "Portal" in self.find_by_test_id("customer-row-cust-003").text
        assert "Print" in self.find_by_test_id("customer-row-cust-004").text
        return self

    def open_inbox(self):
        self.click_by_test_id("nav-inbox")
        self.wait_for_test_id("inbox-page")
        return self

    def expect_inbox_empty_state(self):
        assert "No inbound messages require action" in self.find_by_test_id("inbox-page").text
        return self

    def click_by_test_id(self, test_id):
        self.wait.until(EC.element_to_be_clickable((By.CSS_SELECTOR, self.selector(test_id)))).click()

    def type_by_test_id(self, test_id, value):
        element = self.wait.until(EC.visibility_of_element_located((By.CSS_SELECTOR, self.selector(test_id))))
        element.clear()
        element.send_keys(value)

    def find_by_test_id(self, test_id):
        return self.wait.until(EC.visibility_of_element_located((By.CSS_SELECTOR, self.selector(test_id))))

    def wait_for_test_id(self, test_id):
        return self.find_by_test_id(test_id)

    def wait_for_text(self, test_id, text):
        self.wait.until(EC.text_to_be_present_in_element((By.CSS_SELECTOR, self.selector(test_id)), text))

    def has_test_id(self, test_id):
        return bool(self.driver.find_elements(By.CSS_SELECTOR, self.selector(test_id)))

    def count_test_id(self, test_id):
        return len(self.driver.find_elements(By.CSS_SELECTOR, self.selector(test_id)))

    @staticmethod
    def selector(test_id):
        return f"[data-testid='{test_id}']"

File: tests/playwright-ts/src/pages/comms-flow-app.ts
import { expect, type Page } from "@playwright/test";

type Username = "comms_manager" | "compliance_reviewer";
type CustomerId = "cust-001" | "cust-002" | "cust-003" | "cust-004";
type ProtectedRoute = "/dashboard" | "/templates" | "/campaigns" | "/archive";

const storageKey = "commsflow-state-v1";

function slugify(value: string): string {
  return value.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/(^-|-$)/g, "");
}

export class CommsFlowApp {
  constructor(private readonly page: Page) {}

  async goto() {
    await this.page.goto("/");
    await expect(this.page.getByTestId("login-page")).toBeVisible();
  }

  async loginAs(username: Username, password = "commsflow123") {
    await this.page.getByTestId("username-input").fill(username);
    await this.page.getByTestId("password-input").fill(password);
    await this.page.getByTestId("login-submit").click();
  }

  async openProtectedArchiveWhileLoggedOut() {
    await this.page.goto("/archive");
    await expect(this.page.getByTestId("login-page")).toBeVisible();
  }

  async expectProtectedRoutesRequireLogin(routes: ProtectedRoute[] = ["/dashboard", "/templates", "/campaigns", "/archive"]) {
    for (const route of routes) {
      await this.page.goto(route);
      await expect(this.page.getByTestId("login-page")).toBeVisible();
    }
  }

  async expectDirectRouteAfterLogoutRequiresLogin(route: ProtectedRoute) {
    await this.logout();
    await this.page.goto(route);
    await expect(this.page.getByTestId("login-page")).toBeVisible();
  }

  async setStoredState(state: unknown) {
    await this.page.goto("/login");
    await this.page.evaluate(
      ({ key, value }) => localStorage.setItem(key, typeof value === "string" ? value : JSON.stringify(value)),
      { key: storageKey, value: state }
    );
  }

  async expectCorruptedStorageFallsBackToLogin() {
    await this.setStoredState("{not-valid-json");
    await this.page.goto("/dashboard");
    await expect(this.page.getByTestId("login-page")).toBeVisible();
  }

  async expectUnknownStoredUserFallsBackToLogin() {
    await this.setStoredState({
      user: {
        username: "security_admin",
        role: "Security Admin",
        permissions: ["create_template", "approve_template", "send_campaign"]
      }
    });
    await this.page.goto("/templates");
    await expect(this.page.getByTestId("login-page")).toBeVisible();
  }

  async expectTamperedManagerPermissionsAreCanonical() {
    await this.setStoredState({
      user: {
        username: "comms_manager",
        role: "Comms Manager",
        permissions: ["create_template", "submit_template", "approve_template", "send_campaign"]
      },
      templates: [{ id: "policy-renewal-notice", name: "Policy Renewal Notice", status: "Pending Approval", version: 1, owner: "Comms Manager" }],
      campaigns: [],
      archive: [],
      audit: []
    });
    await this.page.goto("/templates");
    await expect(this.page.getByTestId("templates-page")).toBeVisible();
    await expect(this.page.getByTestId("current-permissions")).toContainText("3 permissions");
    await expect(this.page.getByTestId("approve-template-policy-renewal-notice")).toHaveCount(0);
    await expect(this.page.getByTestId("manager-approval-blocked-policy-renewal-notice")).toContainText("Approval is restricted");
  }

  async attemptInvalidLogin() {
    await this.page.getByTestId("username-input").fill("unknown_user");
    await this.page.getByTestId("password-input").fill("wrong-password");
    await this.page.getByTestId("login-submit").click();
    await expect(this.page.getByTestId("login-error")).toContainText("Invalid credentials");
    await expect(this.page.getByTestId("login-page")).toBeVisible();
  }

  async expectManagerLandsOnDashboard() {
    await expect(this.page.getByTestId("dashboard-page")).toBeVisible();
    await expect(this.page.getByTestId("current-user")).toContainText("Comms Manager");
  }

  async expectReviewerLandsOnTemplates() {
    await expect(this.page.getByTestId("templates-page")).toBeVisible();
    await expect(this.page.getByTestId("current-user")).toContainText("Compliance Reviewer");
  }

  async expectSessionPersistsAfterRefresh() {
    await this.page.reload();
    await expect(this.page.getByTestId("dashboard-page")).toBeVisible();
    await expect(this.page.getByTestId("current-user")).toContainText("Comms Manager");
  }

  async expectManagerAccessProfile() {
    await expect(this.page.getByTestId("access-profile")).toContainText("Comms Manager");
    await expect(this.page.getByTestId("permission-list")).toContainText("Create templates");
    await expect(this.page.getByTestId("permission-list")).toContainText("Submit for approval");
    await expect(this.page.getByTestId("permission-list")).toContainText("Send campaigns");
    await expect(this.page.getByTestId("current-permissions")).toContainText("3 permissions");
  }

  async expectReviewerAccessProfile() {
    await this.openDashboard();
    await expect(this.page.getByTestId("access-profile")).toContainText("Compliance Reviewer");
    await expect(this.page.getByTestId("permission-list")).toContainText("Approve regulated notices");
    await expect(this.page.getByTestId("current-permissions")).toContainText("1 permissions");
  }

  async logout() {
    await this.page.getByTestId("logout-button").click();
    await expect(this.page.getByTestId("login-page")).toBeVisible();
  }

  async expectTopbarNavigation() {
    await this.openDashboard();
    await this.openCustomers();
    await this.openTemplates();
    await this.openCampaigns();
    await this.openInbox();
    await this.openArchive();
  }

  async openTemplates() {
    await this.page.getByTestId("nav-templates").click();
    await expect(this.page.getByTestId("templates-page")).toBeVisible();
  }

  async createTemplate(name = "Policy Renewal Notice") {
    await this.page.getByTestId("template-name-input").fill(name);
    await this.page.getByTestId("template-create-button").click();
    await expect(this.page.getByTestId("template-card-policy-renewal-notice")).toBeVisible();
  }

  async createTemplateWithName(name: string) {
    const id = slugify(name);
    await this.page.getByTestId("template-name-input").fill(name);
    await this.page.getByTestId("template-create-button").click();
    await expect(this.page.getByTestId(`template-card-${id}`)).toBeVisible();
    return id;
  }

  async expectTemplateNamePayloadIsRenderedAsText(payload: string) {
    await this.trackAlertCalls();
    const id = await this.createTemplateWithName(payload);
    await expect(this.page.getByTestId(`template-card-${id}`)).toContainText(payload);
    await this.expectNoScriptPayloadExecuted();
  }

  async expectTemplateDraftMetadata() {
    const card = this.page.getByTestId("template-card-policy-renewal-notice");
    await expect(card).toContainText("Version 1");
    await expect(card).toContainText("Owner: Comms Manager");
    await expect(this.page.getByTestId("template-status-policy-renewal-notice")).toContainText("Draft");
  }

  async expectDuplicateTemplateIsIgnored() {
    await this.createTemplate();
    await expect(this.page.getByTestId("template-card-policy-renewal-notice")).toHaveCount(1);
  }

  async expectTemplatesEmptyState() {
    await expect(this.page.getByTestId("templates-empty-state")).toContainText("No communication templates");
  }

  async expectTemplateDetails() {
    await this.page.getByTestId("template-details-policy-renewal-notice").click();
    await expect(this.page.getByTestId("template-details-policy-renewal-notice")).toContainText("Compliance approval required");
    await expect(this.page.getByTestId("template-details-policy-renewal-notice")).toContainText("Email, SMS, Portal, Print");
  }

  async submitTemplateForApproval() {
    await this.page.getByTestId("submit-template-policy-renewal-notice").click();
    await expect(this.page.getByTestId("template-status-policy-renewal-notice")).toContainText("Pending Approval");
  }

  async expectManagerCannotApprovePendingTemplate() {
    await expect(this.page.getByTestId("manager-approval-blocked-policy-renewal-notice")).toContainText("Approval is restricted");
    await expect(this.page.getByTestId("approve-template-policy-renewal-notice")).toHaveCount(0);
  }

  async expectReviewerCannotCreateTemplate() {
    await expect(this.page.getByTestId("template-permission-note")).toContainText("Template creation restricted");
    await expect(this.page.getByTestId("template-form")).toHaveCount(0);
  }

  async approveTemplate() {
    await this.page.getByTestId("approve-template-policy-renewal-notice").click();
    await expect(this.page.getByTestId("template-status-policy-renewal-notice")).toContainText("Approved");
  }

  async expectApproveButtonIsHiddenAfterApproval() {
    await expect(this.page.getByTestId("approve-template-policy-renewal-notice")).toHaveCount(0);
  }

  async openCampaigns() {
    await this.page.getByTestId("nav-campaigns").click();
    await expect(this.page.getByTestId("campaigns-page")).toBeVisible();
  }

  async expectCampaignRequiresApprovedTemplate() {
    await expect(this.page.getByTestId("campaign-no-template-warning")).toContainText("approved template is required");
    await expect(this.page.getByTestId("campaign-send-button")).toBeDisabled();
  }

  async expectReviewerCannotSendCampaign() {
    await expect(this.page.getByTestId("campaign-permission-warning")).toContainText("Only Comms Managers can send campaigns");
    await expect(this.page.getByTestId("campaign-send-button")).toBeDisabled();
  }

  async expectCampaignsEmptyState() {
    await expect(this.page.getByTestId("campaigns-empty-state")).toContainText("No campaigns");
  }

  async sendPolicyRenewalCampaign() {
    await this.sendPolicyRenewalCampaignToCustomers(["cust-001", "cust-002", "cust-003", "cust-004"]);
    await this.expectCampaignRecipients(4);
  }

  async sendPolicyRenewalCampaignToCustomers(customerIds: CustomerId[]) {
    await this.page.getByTestId("campaign-template-select").selectOption("policy-renewal-notice");
    for (const customerId of ["cust-001", "cust-002", "cust-003", "cust-004"] as CustomerId[]) {
      const checkbox = this.page.getByTestId(`customer-checkbox-${customerId}`);
      const shouldBeChecked = customerIds.includes(customerId);
      if (await checkbox.isChecked() !== shouldBeChecked) {
        await checkbox.click();
      }
    }
    await this.page.getByTestId("campaign-send-button").click();
    await expect(this.page.getByTestId("campaign-card-policy-renewal-may-2026")).toBeVisible();
    await expect(this.page.getByTestId("campaign-status-policy-renewal-may-2026")).toContainText("Sent");
  }

  async sendCampaignWithName(name: string) {
    const id = slugify(name);
    await this.trackAlertCalls();
    await this.page.getByTestId("campaign-name-input").fill(name);
    await this.page.getByTestId("campaign-template-select").selectOption("policy-renewal-notice");
    await this.page.getByTestId("campaign-send-button").click();
    await expect(this.page.getByTestId(`campaign-card-${id}`)).toContainText(name);
    await this.expectNoScriptPayloadExecuted();
  }

  async attemptCampaignWithoutRecipients() {
    await this.page.getByTestId("campaign-template-select").selectOption("policy-renewal-notice");
    for (const customerId of ["cust-001", "cust-002", "cust-003", "cust-004"] as CustomerId[]) {
      const checkbox = this.page.getByTestId(`customer-checkbox-${customerId}`);
      if (await checkbox.isChecked()) {
        await checkbox.click();
      }
    }
    await this.page.getByTestId("campaign-send-button").click();
    await expect(this.page.getByTestId("campaigns-empty-state")).toContainText("No campaigns");
  }

  async expectCampaignRecipients(count: number) {
    await expect(this.page.getByTestId("campaign-card-policy-renewal-may-2026")).toContainText(`Recipients: ${count}`);
  }

  async expectCampaignDetails() {
    await this.page.getByTestId("campaign-details-policy-renewal-may-2026").click();
    await expect(this.page.getByTestId("campaign-details-policy-renewal-may-2026")).toContainText("Customer preference based routing");
    await expect(this.page.getByTestId("campaign-details-policy-renewal-may-2026")).toContainText("Archive records");
  }

  async openArchive() {
    await this.page.getByTestId("nav-archive").click();
    await expect(this.page.getByTestId("archive-page")).toBeVisible();
  }

  async expectArchiveEvidence() {
    await expect(this.page.getByTestId("archive-record-avery-brooks")).toContainText("Email");
    await expect(this.page.getByTestId("archive-record-maya-chen")).toContainText("SMS");
    await expect(this.page.getByTestId("archive-record-nora-singh")).toContainText("Portal");
    await expect(this.page.getByTestId("archive-record-elliot-ward")).toContainText("Print");
  }

  async expectArchiveEvidenceForSelectedCustomers() {
    await expect(this.page.getByTestId("archive-record-avery-brooks")).toContainText("Email");
    await expect(this.page.getByTestId("archive-record-maya-chen")).toContainText("SMS");
    await expect(this.page.getByTestId("archive-record-nora-singh")).toHaveCount(0);
    await expect(this.page.getByTestId("archive-record-elliot-ward")).toHaveCount(0);
  }

  async expectArchiveInitialEmptyState() {
    await expect(this.page.getByTestId("archive-initial-empty-state")).toContainText("No communication evidence");
  }

  async expectArchiveSearchWorks() {
    await this.page.getByTestId("archive-search-input").fill("Maya");
    await expect(this.page.getByTestId("archive-record-maya-chen")).toBeVisible();
    await expect(this.page.getByTestId("archive-record-avery-brooks")).toHaveCount(0);

    await this.page.getByTestId("archive-search-input").fill("FAX");
    await expect(this.page.getByTestId("archive-empty-state")).toContainText("No archive records");
  }

  async expectArchiveSearchByPolicyNumber() {
    await this.page.getByTestId("archive-search-input").fill("POL-33810");
    await expect(this.page.getByTestId("archive-record-nora-singh")).toContainText("Portal");
    await expect(this.page.getByTestId("archive-record-maya-chen")).toHaveCount(0);
  }

  async expectArchiveSearchByCampaignChannelStatusAndTrimmedCase() {
    await this.page.getByTestId("archive-search-input").fill("policy renewal may");
    await expect(this.page.getByTestId("archive-record-avery-brooks")).toBeVisible();
    await expect(this.page.getByTestId("archive-record-elliot-ward")).toBeVisible();

    await this.page.getByTestId("archive-search-input").fill("  portal  ");
    await expect(this.page.getByTestId("archive-record-nora-singh")).toContainText("Portal");
    await expect(this.page.getByTestId("archive-record-maya-chen")).toHaveCount(0);
    await expect(this.page.getByTestId("channel-summary")).toContainText("Portal: 1");

    await this.page.getByTestId("archive-search-input").fill("sent");
    await expect(this.page.getByTestId("channel-summary")).toContainText("Email: 1");
    await expect(this.page.getByTestId("channel-summary")).toContainText("SMS: 1");
    await expect(this.page.getByTestId("channel-summary")).toContainText("Portal: 1");
    await expect(this.page.getByTestId("channel-summary")).toContainText("Print: 1");

    await this.page.getByTestId("archive-search-input").fill("");
    await this.expectArchiveEvidence();
  }

  async expectArchiveSearchPayloadIsSafe(payload: string) {
    await this.trackAlertCalls();
    await this.page.getByTestId("archive-search-input").fill(payload);
    await expect(this.page.getByTestId("archive-empty-state")).toContainText("No archive records");
    await this.expectNoScriptPayloadExecuted();
  }

  async openDashboard() {
    await this.page.getByTestId("nav-dashboard").click();
    await expect(this.page.getByTestId("dashboard-page")).toBeVisible();
  }

  async expectDashboardMetrics(pending: number, approved: number, sent: number, archive: number) {
    await expect(this.page.getByTestId("metric-pending-templates")).toContainText(String(pending));
    await expect(this.page.getByTestId("metric-approved-templates")).toContainText(String(approved));
    await expect(this.page.getByTestId("metric-sent-campaigns")).toContainText(String(sent));
    await expect(this.page.getByTestId("metric-archive-records")).toContainText(String(archive));
  }

  async expectDashboardEvidence() {
    await expect(this.page.getByTestId("metric-approved-templates")).toContainText("1");
    await expect(this.page.getByTestId("metric-sent-campaigns")).toContainText("1");
    await expect(this.page.getByTestId("metric-archive-records")).toContainText("4");
    await expect(this.page.getByTestId("audit-trail")).toContainText("Campaign Policy Renewal May 2026 sent to 4 customers");
  }

  async expectPasswordIsNotExposedAfterLogin() {
    for (const openPage of [() => this.openDashboard(), () => this.openTemplates(), () => this.openCampaigns(), () => this.openArchive()]) {
      await openPage();
      await expect(this.page.getByTestId("app-shell")).not.toContainText("commsflow123");
    }

    const storedState = await this.page.evaluate((key) => localStorage.getItem(key) ?? "", storageKey);
    expect(storedState).not.toContain("commsflow123");
  }

  async expectAuditTrailContainsWorkflowInNewestFirstOrder() {
    const audit = this.page.getByTestId("audit-trail");
    await expect(audit).toContainText("Comms Manager signed in");
    await expect(audit).toContainText("Template Policy Renewal Notice created as draft");
    await expect(audit).toContainText("Template Policy Renewal Notice moved to Pending Approval");
    await expect(audit).toContainText("Template Policy Renewal Notice moved to Approved");
    await expect(audit).toContainText("Campaign Policy Renewal May 2026 sent to 4 customers");

    const text = await audit.innerText();
    expect(text.indexOf("Campaign Policy Renewal May 2026 sent to 4 customers")).toBeLessThan(text.indexOf("Template Policy Renewal Notice moved to Approved"));
    expect(text.indexOf("Template Policy Renewal Notice moved to Approved")).toBeLessThan(text.indexOf("Template Policy Renewal Notice moved to Pending Approval"));
  }

  async openCustomers() {
    await this.page.getByTestId("nav-customers").click();
    await expect(this.page.getByTestId("customers-page")).toBeVisible();
  }

  async expectCustomersPageData() {
    await expect(this.page.getByTestId("customer-row-cust-001")).toContainText("Avery Brooks");
    await expect(this.page.getByTestId("customer-row-cust-001")).toContainText("POL-10491");
    await expect(this.page.getByTestId("customer-row-cust-001")).toContainText("Email");
    await expect(this.page.getByTestId("customer-row-cust-002")).toContainText("SMS");
    await expect(this.page.getByTestId("customer-row-cust-003")).toContainText("Portal");
    await expect(this.page.getByTestId("customer-row-cust-004")).toContainText("Print");
  }

  async openInbox() {
    await this.page.getByTestId("nav-inbox").click();
    await expect(this.page.getByTestId("inbox-page")).toBeVisible();
  }

  async expectInboxEmptyState() {
    await expect(this.page.getByTestId("inbox-page")).toContainText("No inbound messages require action");
  }

  private async trackAlertCalls() {
    await this.page.evaluate(() => {
      (window as unknown as { __commsflowAlerts: string[] }).__commsflowAlerts = [];
      window.alert = (message?: string) => {
        (window as unknown as { __commsflowAlerts: string[] }).__commsflowAlerts.push(message ?? "");
      };
    });
  }

  private async expectNoScriptPayloadExecuted() {
    await expect(this.page.locator("[onerror], [onload]")).toHaveCount(0);
    const alertCount = await this.page.evaluate(() => (window as unknown as { __commsflowAlerts?: string[] }).__commsflowAlerts?.length ?? 0);
    expect(alertCount).toBe(0);
  }
}
